### PR TITLE
ext_proc http bytes tracking support

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
This is a follow up PR of: https://github.com/envoyproxy/envoy/pull/35740.

This PR added the bytes tracking support for ext_proc http service, which basically is counting how many bytes are sent to and received from ext_proc HTTP service.

It takes the stream_info from the active HTTP request, and saved in the ext_proc filter. When the external processing completes, the bytes counters from the stream_info are retrieved and stored into the filter state.

The existing logging test filter is leveraged to test the byte counters in the filter state are correctly updated.

### CODE REVIEW

**Comment by yanjunxiang-google** [comment]
Kind Ping!

**Comment by yanjunxiang-google** [comment]
Kind ping!

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 1]
```diff
@@ -1078,21 +1078,32 @@ void Filter::sendTrailers(ProcessorState& state, const Http::HeaderMap& trailers
...
+  // Only set upstream host in logging info once.
+  if (logging_info_->upstreamHost() == nullptr) {
+    logging_info_->setUpstreamHost(stream_info->upstreamInfo()->upstreamHost());
+  }
+}
+
+void Filter::logStreamInfo() {
   if (!config().grpcService().has_value()) {
+    // HTTP service
+    logStreamInfoBase(stream_info_http_);
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:1100]: Do we need to cache `stream_info_http_` and `setStreamInfo` function?

Can we get access to this through existing members and active_requests_, similar to  gRPC?

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 2]
```diff
@@ -58,7 +53,12 @@ void ExtProcHttpClient::sendRequest(envoy::service::ext_proc::v3::ProcessingRequ
     const auto thread_local_cluster = context().clusterManager().getThreadLocalCluster(cluster);
     if (thread_local_cluster) {
       active_request_ =
-          thread_local_cluster->httpAsyncClient().send(std::move(message), *this, options);
+          thread_local_cluster->httpAsyncClient().startRequest(std::move(headers), *this, options);
+      if (active_request_ != nullptr) {
+        Buffer::OwnedImpl body(req_in_json.value());
+        active_request_->sendData(body, true);
+        callbacks_->setStreamInfo(&active_request_->streamInfo());
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:60]: This PR is missing setting the `upstreamClusterInfo()` in logging_info

**Review by tyxia — COMMENTED** [review]
Thanks for working on this.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 2 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc]: Good catch! done

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 1 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: Sure. We can add a new interface: client->getStreamInfo() to retrieve streamInfo() from active_request_.

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 3]
```diff
@@ -455,11 +455,6 @@ Filter::StreamOpenState Filter::openStream() {
 
     stream_ = config_->threadLocalStreamManager().store(std::move(stream_object), config_->stats(),
                                                         config_->deferredCloseTimeout());
-    // For custom access logging purposes. Applicable only for Envoy gRPC as Google gRPC does not
-    // have a proper implementation of streamInfo.
-    if (grpc_service_.has_envoy_grpc() && logging_info_ != nullptr) {
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:460]: I see you moved the `setClusterInfo(stream_->streamInfo().upstreamClusterInfo())` to logStreamInfoBase, along with setUpstreamHost

It makes sense from ext_proc and good coding practice perspective, but there is the historical reason from logging consumer side that why we want to log the cluster info as earlier as possible.

I will start a thread with internal logging consumer to check this requirement is still true.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:460]: Thanks for the comments!

I thought the logging will only be retried when the call out finished. So, moving the cluster info setup at the end of the call out should be sufficient.  If that's not the case, we should document it somewhere in the code.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:460]: @tyxia It looks like the concern of logging cluster info earlier is cleared.  PTAL the PR again.

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:460]: sgtm. Thanks for following up!

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 4]
```diff
@@ -110,18 +109,24 @@ void ExtProcHttpClient::onFailure(const Http::AsyncClient::Request&,
...
+const Envoy::StreamInfo::StreamInfo* ExtProcHttpClient::getStreamInfo() const {
+  if (active_request_ != nullptr) {
+    return &active_request_->streamInfo();
+  } else {
+    return nullptr;
+  }
+}
+
 void ExtProcHttpClient::onError() {
-  // Cancel if the request is active.
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:118]: Looks like there is behavior change in next revision. Can you elaborate  we are removing this line? And still keep cancel() in the code base.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 5]
```diff
@@ -110,18 +109,24 @@ void ExtProcHttpClient::onFailure(const Http::AsyncClient::Request&,
...
-  cancel();
   ENVOY_LOG(error, "ext_proc HTTP client error condition happens.");
   if (callbacks_) {
     callbacks_->onError();
     callbacks_ = nullptr;
   }
+  active_request_ = nullptr;
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:129]: Compared with previous code, it seems that you are missing active_request_->cancel();? 

why `cancel()` is removed from this function

**Review by tyxia — COMMENTED** [review]
/wait

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 4 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:118]: That's a good question. 

However, there is no logic change here.  The cancel() call is dead code here as from the context, the onError() is always called after active_request_ is set to nullptr. And with active_request_ == nullptr, cancel() is no-op.  

BTW, cancel() call is only needed in the context when the ext_proc filter is destroyed: https://github.com/envoyproxy/envoy/blob/850c7f10cbf5e1132de4eed6b5131d2928ddb95f/source/extensions/filters/http/ext_proc/ext_proc.cc#L474

Another thing is that this PR is trying to move the "active_request_ = nullptr" to after callbacks_->onComplete() and callbacks_->onError() call finished, as these calls will use active_request_ to get the stream_info.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 5 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:129]: Please check comments above.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 4 above)
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:118]: OK then original ext_proc http code has redundancy that is cleaned up here

**Review by tyxia — APPROVED** [review]
LGTM

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**